### PR TITLE
Support for musl (issue #53)

### PR DESCRIPTION
**Pull request by rhencke**
done-ish

### CODE REVIEW

**Review by cemeyer — COMMENTED** [review]
Edit: derp.

File: src/struct/hash.h [hunk 1]
```diff
@@ -15,16 +15,17 @@
  */
 
 #include <stdint.h>
+#include <asm/bitsperlong.h>
```
Comment (cemeyer) [inline src/struct/hash.h:18]: This header is Linux-specific and not portable.

File: src/struct/hash.h [hunk 2]
```diff
@@ -15,16 +15,17 @@
...
 
 /* 2^31 + 2^29 - 2^25 + 2^22 - 2^19 - 2^16 + 1 */
 #define GOLDEN_RATIO_PRIME_32 0x9e370001UL
 /*  2^63 + 2^61 - 2^57 + 2^54 - 2^51 - 2^18 + 1 */
 #define GOLDEN_RATIO_PRIME_64 0x9e37fffffffc0001UL
 
-#if __WORDSIZE == 32
+#if __BITS_PER_LONG == 32
```
Comment (cemeyer) [inline src/struct/hash.h:25]: I'd suggest `__LP64__` and `__ILP32__` per https://sourceforge.net/p/predef/wiki/DataModels/ .

**Review by cemeyer — CHANGES_REQUESTED** [review]

File: src/struct/dutil.h [hunk 3]
```diff
@@ -41,7 +41,7 @@ static inline __attribute__((const)) bool is_power_of_2(unsigned long n)
  * This is defined the same way as ffs.
  * Note fls(0) = 0, fls(1) = 1, fls(0x80000000) = 32.
  */
-static __always_inline int fls(int x)
+static inline __attribute__((__always_inline__)) int fls(int x)
```
Comment (cemeyer) [inline src/struct/dutil.h:44]: Suggest instead:

```
#ifndef __always_inline
#define __always_inline __attribute__((__always_inline__))
#endif
```

File: src/struct/hash.h (hunk 1 above)
Comment (rhencke) [reply src/struct/hash.h:18]: I thought so too until I saw [this](https://github.com/taviso/ctypes.sh/blob/091e64421507c3502b38b1b88d78b12b4269e80e/lib/asm/bitsperlong.h)... :)

Comment (cemeyer) [reply src/struct/hash.h:18]: Aha.  Well, I'd still suggest `__LP64__` and `__ILP32__`.  Unlike the stub asm/bitsperlong.h, they work with 32- and 64-bit architectures other than x86.  Tavis might have a different opinion.

Comment (rhencke) [reply src/struct/hash.h:18]: I'll take a shot at changing this to use __ILP32__ and __LP64__ and cleaning this up throughout.  Thanks. :)

File: src/struct/hash.h [hunk 4]
```diff
@@ -21,10 +21,10 @@
 /*  2^63 + 2^61 - 2^57 + 2^54 - 2^51 - 2^18 + 1 */
 #define GOLDEN_RATIO_PRIME_64 0x9e37fffffffc0001UL
 
-#if __WORDSIZE == 32
+#if defined(__ILP32__)
 #define GOLDEN_RATIO_PRIME GOLDEN_RATIO_PRIME_32
 #define hash_long(val, bits) hash_32(val, bits)
-#elif __WORDSIZE == 64
+#elif defined(__LP64__)
```
Comment (rhencke) [inline src/struct/hash.h:27]: Now using `__ILP32__` and `__LP64__`.

**Review by rhencke — COMMENTED** [review]
As far as I could tell, the code removed was unused, and removing it was easier than fixing it. :)

**Review by cemeyer — APPROVED** [review]
LGTM.  Is that all that prevented it working on musl?  Thanks!

**Comment by rhencke** [comment]
Hi @cemeyer !  Thank you for the review!

Not yet.. more to come, stay tuned.. (unfortunately past couple days has been Java heap dump analysis which isn't nearly a fun..)

**Comment by rhencke** [comment]
It builds now! :D  I'll let you know if it works!

**Comment by rhencke** [comment]
Closer!
```
scar:~$ source /usr/local/bin/ctypes.sh
bash: enable: cannot open shared object /usr/local/lib/ctypes.so: Error relocating /usr/local/lib/ctypes.so: _obstack_newchunk: symbol not found
can't find the ctypes.so library, run make install?
```
Looks like figuring out `_obstack_newchunk` is next.

**Comment by rhencke** [comment]
Hi @cemeyer,

I have been looking at obstacks - they are new to me.  They appear to be a [glibc extension](https://www.gnu.org/software/libc/manual/html_node/Obstacks.html) that provides a stack data structure as a contiguous slice of memory, allowing for variable-size allocations for each object, and also supporting dynamically _extending_ the current (top of stack) object's allocation without a fixed limit with the caveat that its pointer wont be stable until the object is finalized.  There's a nontrivial amount of functionality it provides.

But.. we only use `obstack_init`, `obstack_alloc` and `obstack_free`.  We don't use any of the dynamic resizing capabilities, or any of that.  Given that, I'm tempted to remove the use of obstack and find a mechanism that doesn't use a glibc-ism.

There's two things I need to be careful about (that I can think of) if I do this:
1) There might be code that assumes these objects are allocated in contiguous memory.
2) If the object passed to `obstack_free` is not at the top of the stack, all objects above that object on the stack are freed as well.  If we are depending on this anywhere, a naïve replacement with, say, `malloc` and `free` might leak memory.

A second approach is possible - there exists a port of just `obstack` to musl called [musl-obstacl](https://github.com/void-linux/musl-obstack), that is provided in many musl-based Linux distributions (mine inlcuded).  I could leave all of this code alone and figure out how to teach autoconf/automake to expect this on musl-based systems.

What do you think?

**Comment by taviso** [comment]
That code is inherited from the dwarves library, which is also used by `pahole`...I wonder what `pahole` does on musl systems?

**Comment by rhencke** [comment]
I'll take a look and find out. :)

**Comment by rhencke** [comment]
It.. uh.. doesn't build. :)  At least from https://git.kernel.org/pub/scm/devel/pahole/pahole.git directly.  It actually breaks with some of the stuff this PR has fixed so far.

**Comment by rhencke** [comment]
Aha.

Ok, so Void Linux fixes this through a series of custom patches to the original source, that are basically doing the same thing I'm doing, so that makes me feel better. :) (though I think the WORDSIZE fix @cemeyer helped with we have a superior fix).

* https://github.com/void-linux/void-packages/blob/master/srcpkgs/pahole/patches/fix_always_inline.patch
* https://github.com/void-linux/void-packages/blob/master/srcpkgs/pahole/patches/fix_sysconf.patch
* https://github.com/void-linux/void-packages/blob/master/srcpkgs/pahole/patches/fix_wordsize.patch

So.. I can take the same approach, and basically I can leave this code alone and just teach automake/autoconf how to (if necessary) link in `obstack` on musl systems.

**Comment by cemeyer** [comment]
You can grab obstack from gnulib: https://github.com/coreutils/gnulib/blob/master/lib/obstack.c

**Comment by rhencke** [comment]
So.. I think I understand things better now.  It looks like all I really needed to do was to make the existing obstack check unconditional.  That causes the existing copy of obstack.c we already have in the lib folder to be used, if the autoconf check for working obstacks fails (which it does for musl, which is fine).  So we basically will just statically link in our copy, if the system doesn't provide it in libc.

And.. with that!
```bash
scar:~/src/github.com/taviso/ctypes.sh$ source /usr/local/bin/ctypes.sh 
scar:~/src/github.com/taviso/ctypes.sh$ dlcall printf $'hello world\n'
hello world
```

**Comment by rhencke** [comment]
Thanks very much @cemeyer and @taviso for the review and guidance!

**Comment by taviso** [comment]
It seems okay to merge to me, do the tests pass with musl?

File: src/struct/dutil.h [hunk 5]
```diff
@@ -38,227 +35,6 @@ static inline __attribute__((const)) bool is_power_of_2(unsigned long n)
         return (n != 0 && ((n & (n - 1)) == 0));
 }
 
-/**
```
[the remark this answers is not in the document]
Comment (cemeyer) [reply src/struct/dutil.h:37]: So this header came from pahole originally; if we want to update to newer versions of pahole, I wonder if making this big change would make that more difficult?  Maybe not.

File: src/struct/dwarves.c [hunk 6]
```diff
@@ -2222,11 +2222,11 @@ void cus__delete(struct cus *cus)
 	free(cus);
 }
 
-void dwarves__fprintf_init(uint16_t user_cacheline_size);
```
Comment (cemeyer) [inline src/struct/dwarves.c:2225]: Again, it seems maybe bad to be making refactoring changes to what is mostly a drop-in file from pahole.  Not a huge deal, but this kind of stuff accumulates eventually.

Comment (rhencke) [reply src/struct/dwarves.c]: I've reverted these changes in the latest version.  It should be much less invasive now.

File: src/struct/dutil.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (rhencke) [reply src/struct/dutil.h:37]: If they changed the function dependencies, that could bite us, yes.  Let me see what Void did - maybe I can get away with something less drastic.

**Comment by rhencke** [comment]
@taviso Not yet, but I don't think the fixes are drastic.  I'll address that next.

**Comment by rhencke** [comment]
Well.. the tests did exactly what they were supposed to, I think - they caught a real difference.

The following code acts differently between musl and glibc, and I'm not sure why:
```bash
dlopen libm.so.6 _INVALID_FLAG
```

On a glibc system, it succeeds, with exit code 0:
```
bash: dlopen: warning: invalid or unrecognised rtld flag ignored: _INVALID_FLAG
0x7f30c7e63b40
```

On musl, it fails, with exit code 1:
```
bash: dlopen: warning: invalid or unrecognised rtld flag ignored: _INVALID_FLAG
bash: dlopen: dlopen("libm.so.6", 0) failed, libm.so.6: invalid mode for dlopen(): Invalid argument
```

It doesn't seem to matter if it's specifically `libm.so.6` - other shared libraries exhibit the same behavior.  Loading _without_ the `_INVALID_FLAG` succeeds on both glibc and musl.

**Comment by taviso** [comment]
Hmm, so `dlopen(foo, 0)` is *accepted* in musl?

I think that's okay, maybe change the test to something like this: 

``diff
diff --git a/test/dlopen.sh b/test/dlopen.sh
index 40aa8f2..b33582f 100644
--- a/test/dlopen.sh
+++ b/test/dlopen.sh
@@ -16,7 +16,7 @@ exec 2> /dev/null

 dlopen _invalid_lib_name_               && failure
 dlopen libm.so.6                        || failure
-dlopen libm.so.6 _INVALID_FLAG          && failure
+dlopen libm.so.6 0xdeadbeef             && failure
 dlopen                                  && failure
 dlopen libm.so.6 RTLD_GLOBAL RTLD_LAZY  || failure
```

I think I just wanted to make sure an invalid mode was handled correctly, and the code ignores invalid symbolic names (maybe that was a bad idea, but some systems support flags that others don't). So `dlopen foo INVALID` ends up being mode 0, which gets rejected.

I mean, it will be an issue in future if 0xdeadbeef is all valid flags, but I think we can cross that bridge when we get to it!

**Comment by rhencke** [comment]
Ah, yes - exactly.  Sorry, I had it backwards, fixed my comment.

So I went to try that and found.. unfortunately.. musl also differs and succeeds if you give it `0xdeadbeef`...

...so, who's right? I looked it up, and [POSIX says](https://pubs.opengroup.org/onlinepubs/009695399/functions/dlopen.html)...
> If neither RTLD_GLOBAL nor RTLD_LOCAL are specified, then the default behavior is unspecified.

Well.. that.. doesn't help.

I can think of a couple options:
1) Introduce a bit of code into ctypes.sh that, if an invalid mode is seen, fail early.  Matches today's behavior in glibc and is a bit more user friendly.
2) Remove the test case, under the logic that it's technically undefined behavior.

I'm leaning towards 1.. I'll give it a try at least.

**Comment by rhencke** [comment]
What I settled on: do what glibc does for `RTLD_LAZY` and `RTLD_NOW`.

(The `RTLD_GLOBAL` or `RTLD_LOCAL` one is tougher because glibc allows much more than that, so I'm punting on that one...)

Next up is `stirfry.sh`.  Getting close! :)

**Comment by rhencke** [comment]
Everything passes now in musl! :)

File: test/sha1.sh [hunk 7]
```diff
@@ -15,8 +15,12 @@ declare s=(uint8:{1..20})
 
 # SHA_CTX is a typedef, not a struct, so you should use -a
 if ! sizeof -am ctx SHA_CTX; then
-    echo FIXME: failed to create SHA_CTX structure
-    exit 1
+    echo 'WARNING: failed to create SHA_CTX structure (missing debug symbols?) - falling back to manual definition'
+    # see SHAstate_st in <sha.h> for structure definition
+    if ! dlcall -r pointer -n ctx malloc $(($(sizeof unsigned) * (5 + 2 + 16 + 1))); then
```
Comment (cemeyer) [inline test/sha1.sh:20]: This seems a bit odd; openssl shouldn't have anything to do with the libc involved.  Any idea what went wrong?

**Review by cemeyer — COMMENTED** [review]
LGTM other than the SHA_CTX thing.

File: test/sha1.sh (hunk 7 above)
Comment (rhencke) [reply test/sha1.sh]: I wasn't sure at first, but eventually I found this which pointed me towards the answer: https://github.com/taviso/ctypes.sh/wiki/Structures#troubleshooting-automatic-structure

Thankfully, my distribution had an equivalent debug symbols package.. I just didn't notice.  It now works fine without this change, so I've removed it.  Thanks. :)

**Review by cemeyer — APPROVED** [review]
LGTM!

**Comment by taviso** [comment]
Thanks @rhencke, merged!

**Comment by rhencke** [comment]
Thank you both for the review!

**Pull request merged**